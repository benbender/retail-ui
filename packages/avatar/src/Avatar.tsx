import { Theme, useThemeCtx } from '@retail-ui/theme'
import { cloneElement } from '@retail-ui/utils'
import clsx from 'clsx'
import * as React from 'react'

type ReactDivProps = React.HTMLAttributes<HTMLDivElement>
type Ref = HTMLDivElement

type AvatarSize = keyof Theme['AvatarStyles']['size']
// type AvatarStatus = keyof Theme['Avatar']['status']

export type AvatarProps = {
  size?: AvatarSize
  // shape:
  // image
  alt?: string
  src?: string
  // text
  title?: string
  // icon
  icon?: React.ReactNode
  // status: AvatarStatus
}

export const Avatar = React.forwardRef<Ref, ReactDivProps & AvatarProps>(
  (props, ref) => {
    const { size = 'base', className, alt, src, title, icon, ...rest } = props
    const {
      theme: { AvatarStyles },
    } = useThemeCtx()

    const sizeCls = AvatarStyles.size

    const cls = clsx(
      className,
      AvatarStyles.base,
      sizeCls[size],
      icon && AvatarStyles.avatarIcon,
    )

    let childrenToRender // priority image => icon => text
    if (src) {
      const imgCls = clsx(AvatarStyles.image.base)
      childrenToRender = (
        <img className={imgCls} src={src} alt={alt} loading="lazy" />
      )
    } else if (icon) {
      const iconSizeCls = AvatarStyles.icon.size
      childrenToRender = cloneElement(icon, { className: iconSizeCls })
    } else {
      // title
      const titleSizeCls = AvatarStyles.title.size
      const titleCls = clsx(AvatarStyles.title.base, titleSizeCls[size])
      childrenToRender = <span className={titleCls}>{title}</span>
    }

    return (
      <div ref={ref} className={cls} {...rest}>
        {childrenToRender}
        {/* Some shadow */}
        <div
          className="absolute inset-0 rounded-full shadow-inner"
          aria-hidden="true"
        ></div>
      </div>
    )
  },
)

Avatar.displayName = 'Avatar'
