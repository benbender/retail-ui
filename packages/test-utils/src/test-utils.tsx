import { defaultTheme, ThemeProvider } from '@retail-ui/theme'
import {
  render as RtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react'
import * as React from 'react'

type UI = Parameters<typeof RtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult => {
  return RtlRender(
    <ThemeProvider value={{ theme: defaultTheme }}>
      <Wrapper>{ui}</Wrapper>
    </ThemeProvider>,
    options,
  )
}

export { fireEvent } from '@testing-library/react'
