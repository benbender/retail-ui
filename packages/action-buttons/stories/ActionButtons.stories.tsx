import * as React from 'react'

import { ActionButtons, ActionButtonsProps } from '../src'

const meta = {
  title: 'Retail-UI/ActionButtons',
  component: ActionButtons,
}

export default meta

const Template = (args: ActionButtonsProps) => (
  <ActionButtons {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
