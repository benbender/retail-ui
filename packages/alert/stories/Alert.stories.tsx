import * as React from 'react'

import { Alert, AlertProps } from '../src'

const meta = {
  title: 'Retail-UI/Alert',
  component: Alert,
}

export default meta

const Template = (args: AlertProps) => (
  <Alert {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'A alert that informs you of stuff',
}
