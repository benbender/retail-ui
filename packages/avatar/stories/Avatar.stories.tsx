import * as React from 'react'

import { Avatar, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/Avatar',
  component: Avatar,
}

export default meta

const Template = (args: AvatarProps) => (
  <Avatar {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  src: `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`,
}
