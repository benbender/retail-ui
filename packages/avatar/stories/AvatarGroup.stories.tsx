import * as React from 'react'

import { Avatar, AvatarGroup, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/AvatarGroup',
  component: AvatarGroup,
}

export default meta

const src = `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`

const Group = (args: AvatarProps) => (
  <AvatarGroup data-testid={meta.title}>
    <Avatar src={src} {...args} />
    <Avatar title="PJ" size="base" {...args} />
    {/* <Avatar icon={UserSolid} {...args} /> */}
  </AvatarGroup>
)

export const Simple = Group.bind({})
// @ts-ignore
Simple.args = {}
