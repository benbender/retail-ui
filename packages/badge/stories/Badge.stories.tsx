import * as React from 'react'

import { Badge, BadgeProps } from '../src'

const meta = {
  title: 'Retail-UI/Badge',
  component: Badge,
}

export default meta

const Template = (args: BadgeProps) => (
  <Badge {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Badge',
}
