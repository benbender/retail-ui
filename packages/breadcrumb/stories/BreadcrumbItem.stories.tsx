import * as React from 'react'

import { Breadcrumb, BreadcrumbItem, BreadcrumbProps } from '../src'

const meta = {
  title: 'Retail-UI/BreadcrumbItem',
  component: BreadcrumbItem,
}

export default meta

const Template = (args: BreadcrumbProps) => (
  <Breadcrumb data-testid={meta.title}>
    <BreadcrumbItem {...args}>
      <a href="#">Dashboard</a>
    </BreadcrumbItem>
    <BreadcrumbItem {...args}>
      <a href="#">Project</a>
    </BreadcrumbItem>
    <BreadcrumbItem isActive {...args}>
      Export
    </BreadcrumbItem>
  </Breadcrumb>
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Breadcrumb',
}
