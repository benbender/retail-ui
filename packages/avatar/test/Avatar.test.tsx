import { render } from '@retail-ui/test-utils'
import * as React from 'react'

import * as Avatar from '../stories/Avatar.stories'

describe('Avatar', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Avatar.Simple />)
    expect(asFragment()).toMatchSnapshot()
  })
})
