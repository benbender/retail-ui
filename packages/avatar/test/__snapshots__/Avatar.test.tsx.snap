// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders correctly 1`] = `
<DocumentFragment>
  <div
    class="relative rounded-full select-none text-center no-underline h-10 w-10 "
  >
    <span
      class="absolute rounded-full inset-0 font-bold uppercase flex items-center justify-center h-full w-full bg-purple-300 text-purple-600 no-underline text-base"
    />
    <div
      aria-hidden="true"
      class="absolute inset-0 rounded-full shadow-inner"
    />
  </div>
</DocumentFragment>
`;
